### PR TITLE
feat(events tracking): add abstract class and logging implementation

### PR DESCRIPTION
**Pull request by victoria-yining-huang**
[design doc](https://www.notion.so/sentry/Conversion-rate-of-ingest-transactions-to-save-trx-1298b10e4b5d801ab517c8e2218d13d5)

need to track the completion of each stage, to 1) compute events conversion rates 2) enable debugging visibility into where events are being dropped

the usage will be heavily sampled to not blow up traffic

this PR only adds REDIS_PUT stage, in subsequent PRs I will add all the other stages listed in EventStageStatus class

**!!!!!IMPORTANT!!!!!!**
hash based sampling
here's a [blog post](https://www.rsyslog.com/doc/tutorials/hash_sampling.html) explaining hash based sampling, which would provide "all or nothing" logging for the events sampled across the entire pipeline. That's the idea I want to implement

the hashing algorithm used must be consistent and uniformly distributed in order for all or nothing sampling to work. 
I cannot find references that say that md5 is consistent and evenly distributed other than various [stackoverflow pages](https://crypto.stackexchange.com/questions/14967/distribution-for-a-subset-of-md5). All the official sources are too academic and long and i can't understand

----------
for reviewers:
please review with the thoughts of how this can be generalized to other pipelines as well, such as errors

### CODE REVIEW

File: src/sentry/utils/event_tracker.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+    redis_put
+
+    save_event_started
+
+    save_event_finished
+
+    snuba_topic_put
+
+    billing_topic_put
```
Comment (fpacifici) [inline src/sentry/utils/event_tracker.py:21]: unneeded.

File: src/sentry/utils/event_tracker.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import logging
+from abc import ABC, abstractmethod
+from enum import Enum
+
+
+class EventStageStatus(Enum):
+    START = "start"
+    END = "end"
```
Comment (fpacifici) [inline src/sentry/utils/event_tracker.py:8]: Consider making the enum an integer. That would take a lot less memory in redis.

The galaxy brain version would be to create a bitmap. One bit per phase. Each event only needs 12 bits to encode all phases. https://redis.io/docs/latest/develop/data-types/bitmaps/

File: src/sentry/utils/event_tracker.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+
+class EventTracker(EventTrackerBackend):
+    """
+    Logger-based implementation of EventTrackerBackend. The data will be saved in BigQuery using Google Log Sink
+    """
+
+    def __init__(self):
+        self.logger = logging.getLogger("EventTracker")
+        logging.basicConfig(level=logging.INFO)
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:59]: doesn't this modify the root logger for all of the application?

File: src/sentry/utils/event_tracker.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        Records how far an event has made it through the ingestion pipeline.
+
+        Args:
+            event_id (str): Unique identifier of the event.
+            status (ProcessingPhase): The status of each step of the data lineage, either START or END.
+        """
+        raise NotImplementedError
+
+
+class EventTracker(EventTrackerBackend):
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:52]: Do you need the base class for any reason? Are you planning to provide a second implementation? I'd just remove it if not and make it a bit simpler

File: src/sentry/utils/event_tracker.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    """
+    Logger-based implementation of EventTrackerBackend. The data will be saved in BigQuery using Google Log Sink
+    """
+
+    def __init__(self):
+        self.logger = logging.getLogger("EventTracker")
+        logging.basicConfig(level=logging.INFO)
+
+    def record_event_stage_status(self, event_id: str, status: EventStageStatus):
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:61]: can you please add sampling logic in here? we'll be in trouble if we log every event

File: src/sentry/utils/event_tracker.py (hunk 2 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py:8]: not writing to redis with the logging implementation but i get the idea

File: src/sentry/utils/event_tracker.py (hunk 1 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: updated the list

File: src/sentry/utils/event_tracker.py (hunk 4 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py:48]: > Are you planning to provide a second implementation?

not anytime soon, for the purpose of tracking event stages logger based implementation should be sufficient. Maybe another metric based implementation can be used for refunding?

Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py:48]: im gonna delete for now

File: src/sentry/utils/event_tracker.py (hunk 5 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py:42]: added, default is 0.01

File: src/sentry/utils/event_tracker.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        self.logger = logging.getLogger("EventTracker")
+        self.sample_rate = sample_rate
+
+    def record_event_stage_status(self, event_id: str, status: EventStageStatus):
+        if random.random() > self.sample_rate:
+            return
+        """
+        Records how far an event has made it through the ingestion pipeline.
+        """
+        self.logger.info(f"EventTracker recorded event {event_id} - {status.value}")
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:51]: if you are building log analysis tools on top of this structured logging might make life easier

File: src/sentry/utils/event_tracker.py (hunk 5 above)
Comment (lynnagara) [reply src/sentry/utils/event_tracker.py:45]: you don't want random though do you? don't you need consistent sampling so that the same events get sampled at every stage of the pipeline?

File: src/sentry/utils/event_tracker.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        return True
+
+    def record_event_stage_status(
+        self, event_id: str, status: EventStageStatus, is_tracked: bool = False
+    ):
+        """
+        Records how far an event has made it through the ingestion pipeline.
+        """
+        if is_tracked:
+            self.logger.info(f"EventTracker recorded event {event_id} - {status.value}")
```
Comment (markstory) [inline src/sentry/utils/event_tracker.py:58]: If you log the stage as an `extra` parameter you'll be able to query it in the logging infra.

Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: ooh done thanks

File: src/sentry/utils/event_tracker.py (hunk 5 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: yeah i forgot, fixed

File: src/sentry/utils/event_tracker.py [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    ppf_topic_put
+
+    post_process_started
+
+    post_process_finished / the same as redis_deleted
+    """
+
+
+class EventTracker:
```
Comment (evanpurkhiser) [inline src/sentry/utils/event_tracker.py:32]: I think you can probably just have a couple functions here, you don't need a class for this

File: src/sentry/utils/event_tracker.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    Logger-based implementation of EventTrackerBackend. The data will be saved in BigQuery using Google Log Sink
+    """
+
+    def __init__(self, sample_rate: float = 0.01):
+        """
+        Args:
+            sample_rate (float): The probability (0.0 to 1.0) that an event is recorded.
+                                 A value of 1.0 records all events, 0.1 records approximately 10% of events.
+        """
+        self.logger = logging.getLogger("EventTracker")
```
Comment (evanpurkhiser) [inline src/sentry/utils/event_tracker.py:43]: You just move this to the module level

File: src/sentry/utils/event_tracker.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+        return True
+
+    def record_event_stage_status(
+        self, event_id: str, status: EventStageStatus, is_tracked: bool = False
+    ):
+        """
+        Records how far an event has made it through the ingestion pipeline.
+        """
+        if is_tracked:
+            extra = {"event_id": event_id, "status": status}
```
Comment (evanpurkhiser) [inline src/sentry/utils/event_tracker.py:58]: You need to do `EventStageStatus.value` to get the string value from the enum.

Alternatively you can use [`StrEnum`](https://docs.python.org/3/library/enum.html#enum.StrEnum) which is probably a better approach

**Review by bmckerry — APPROVED** [review]
lgtm but should probably get a review from someone more familiar with this kind of sampling setup

File: src/sentry/utils/event_tracker.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+
+def is_sampled_to_track(event_id: str, sample_rate) -> bool:
+    """
+    Normalize the integer to a float in the range [0, 1)
+    The md5 hashing algorithm is consistent and will make sure the same event_id's are sampled,
+    and provide all or nothing logging.
+    """
+    hash_float = int(md5_text(event_id).hexdigest(), 16) / (2**128 - 1)
+    if hash_float < sample_rate:
```
Comment (mjq) [inline src/sentry/utils/event_tracker.py:38]: As discussed on Slack: the max value of the numerator in the hash_float calculation is the same as the denominator, so hash_float has a range of [0, 1]. I'd recommend using `<=` for the sample rate comparison - in the infinitesimally small case of exactly 0 or 1, I'd rather incorrectly log at sample rate `0` than miss an event at sample rate `1.0`.

```suggestion
    Normalize the integer to a float in the range [0, 1]
    The md5 hashing algorithm is consistent and will make sure the same event_id's are sampled,
    and provide all or nothing logging.
    """
    hash_float = int(md5_text(event_id).hexdigest(), 16) / (2**128 - 1)
    if hash_float <= sample_rate:
```

Comment (mjq) [reply src/sentry/utils/event_tracker.py:38]: Alternatively to avoid the edge case (but I don't think it really matters)

```suggestion
    Normalize the integer to a float in the range [0, 1]
    The md5 hashing algorithm is consistent and will make sure the same event_id's are sampled,
    and provide all or nothing logging.
    """
    hash_float = int(md5_text(event_id).hexdigest(), 16) / (2**128 - 1)
    if sample_rate > 0 && hash_float <= sample_rate:
```

File: src/sentry/utils/event_tracker.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+    snuba_topic_put
+
+    commit_log_topic_put
+
+    ppf_topic_put
+
+    post_process_started
+
+    post_process_finished / the same as redis_deleted
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:24]: What does this mean? Mind adding them in the EventStageStatus enum rather than in the comment if that's what it's supposed to represent. How come redis put is 1, does that mean it's the first step in the pipeline that will be recorded?

File: src/sentry/utils/event_tracker.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+logger = logging.getLogger("EventTracker")
+
+
+def is_sampled_to_track(event_id: str, sample_rate) -> bool:
+    """
+    Normalize the integer to a float in the range [0, 1)
+    The md5 hashing algorithm is consistent and will make sure the same event_id's are sampled,
+    and provide all or nothing logging.
+    """
+    hash_float = int(md5_text(event_id).hexdigest(), 16) / (2**128 - 1)
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:37]: I'd just use a faster non cryptographic hash function here like `hash(event_id)` here

File: src/sentry/utils/event_tracker.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    """
+    Records how far an event has made it through the ingestion pipeline.
+    """
+    if is_sampled_to_track(event_id, sample_rate):
+        extra = {"event_id": event_id, "status": status}
+        _do_record(extra)
+
+
+def _do_record(extra):
+    logger.info("EventTracker.recorded", extra=extra)
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:55]: nit, but i'd just inline this. is there any value in having a separate function for this one line? i feel like it just adds some indirection for no reason

File: src/sentry/utils/event_tracker.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    post_process_started
+
+    post_process_finished / the same as redis_deleted
+    """
+
+
+logger = logging.getLogger("EventTracker")
+
+
+def is_sampled_to_track(event_id: str, sample_rate) -> bool:
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:31]: function name could be more straightforward

```suggestion
def should_track(event_id: str, sample_rate) -> bool:
```

File: src/sentry/utils/event_tracker.py (hunk 14 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py:55]: i made it a separate function so in tests i can verify it's called once or not called

File: tests/sentry/ingest/ingest_consumer/test_ingest_consumer_processing.py [hunk 16]
```diff
@@ -161,6 +161,86 @@ def test_transactions_spawn_save_event_transaction(
     )
 
 
+@django_db_all
+def test_transactions_save_event_transaction_is_tracked(default_project):
```
Comment (lynnagara) [inline tests/sentry/ingest/ingest_consumer/test_ingest_consumer_processing.py:165]: What are these tests actually testing? That true just logs a line and false does not. Is that actually worthwhile, especially if the sampling function is mocked anyway?

Comment (victoria-yining-huang) [reply tests/sentry/ingest/ingest_consumer/test_ingest_consumer_processing.py:165]: i think still worth? Because it's testing in the `process_event` function, logging is getting called or not for a transaction

File: tests/sentry/ingest/ingest_consumer/test_ingest_consumer_processing.py [hunk 17]
```diff
@@ -161,6 +161,86 @@ def test_transactions_spawn_save_event_transaction(
     )
 
 
+@django_db_all
```
Comment (lynnagara) [inline tests/sentry/ingest/ingest_consumer/test_ingest_consumer_processing.py:164]: you should probably write unit tests for `event_tracker.py`. i don't think you need tests for this in every ingest consumer, especially if this is supposed to be extended and used in a bunch of places later.

Comment (victoria-yining-huang) [reply tests/sentry/ingest/ingest_consumer/test_ingest_consumer_processing.py:164]: if it's just unit tests for `event_tracker.py` what should the tests be testing? Wouldn't that just be is logging called or not?

File: tests/sentry/ingest/ingest_consumer/test_ingest_consumer_processing.py (hunk 16 above)
Comment (lynnagara) [reply tests/sentry/ingest/ingest_consumer/test_ingest_consumer_processing.py:165]: I think you should make `event_tracker` generic so it can be used in all ingest pipelines, and tests for the `event_tracker`.

It shouldn't be the responsibility of the pipelines that will use this module to write the tests.

File: src/sentry/utils/event_tracker.py (hunk 11 above)
Comment (fpacifici) [reply src/sentry/utils/event_tracker.py:38]: Please do not rely on sampling = 0.0 to let any event pass. That would also be your killswitch. 
If anything you should do one of theese two:
- checked whether sampling is 0.0 and quit immediately before hashing the event id.
- have another option to serve as a killswitch so that if it is off you do not do anything inside the method including hashing.

File: src/sentry/utils/event_tracker.py
Comment (fpacifici) [inline src/sentry/utils/event_tracker.py:1]: Please write a unit test for module. 
The integration test is not testing the the sampling logic as it is mocking it.

File: src/sentry/utils/event_tracker.py (hunk 13 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: resolved on slack, hash() is not consistent across different python processes

File: src/sentry/utils/event_tracker.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: added, also changed class name to `TransactionStageStatus` to be less specific, so other pipelines can have their own enums such as `ErrorStageStatus`

File: src/sentry/utils/event_tracker.py (hunk 11 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: added if sample rate ==0 return

File: src/sentry/utils/event_tracker.py [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+def record_sampled_event_stage_status(
+    event_id: str, status: EventStageStatus, sample_rate: float = 0.01
+) -> None:
+    """
+    Records how far an event has made it through the ingestion pipeline.
+    """
+    if is_sampled_to_track(event_id, sample_rate):
+        extra = {"event_id": event_id, "status": status}
+        _do_record(extra)
```
[the remark this answers is not in the document]
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: > Also, since you are planning to use this for multiple events types, please have a structured option where you can set different sampling rates per event type.

this is good idea but i think can be iterated upon? I want to keep this one PR just for transactions and turn it on soon

File: tests/sentry/utils/test_event_tracker.py
Comment (victoria-yining-huang) [inline tests/sentry/utils/test_event_tracker.py:1]: @fpacifici added unit test just for the event_tracker file, removed testing it in the ingest consumer tests

File: src/sentry/utils/event_tracker.py
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py:1]: done

File: tests/sentry/ingest/ingest_consumer/test_ingest_consumer_processing.py (hunk 16 above)
Comment (victoria-yining-huang) [reply tests/sentry/ingest/ingest_consumer/test_ingest_consumer_processing.py]: done

File: src/sentry/utils/event_tracker.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: added TODO comment

File: src/sentry/utils/event_tracker.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: using options now

File: src/sentry/ingest/consumer/processors.py [hunk 19]
```diff
@@ -202,6 +203,10 @@ def process_event(
         else:
             with metrics.timer("ingest_consumer._store_event"):
                 cache_key = processing_store.store(data)
+            track_sampled_event(
+                data["event_id"], data.get("type"), TransactionStageStatus.REDIS_PUT
```
Comment (lynnagara) [inline src/sentry/ingest/consumer/processors.py:207]: I think you want the pipeline name not `data.get("type")` here. They are not always the same.

File: src/sentry/utils/event_tracker.py [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    Do NOT change the int values of existing items. It will change meaning of existing tracking data
+    This class is DELETE or APPEND ONLY
+    """
+
+    # the transaction is stored to rc-transactions
+    REDIS_PUT = 1
+
+    # a save_transactions task is kicked off
+    SAVE_TRX_STARTED = 2
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:22]:
```suggestion
    SAVE_TXN_STARTED = 2
```

File: src/sentry/utils/event_tracker.py [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+
+    # a save_transactions task is finished
+    SAVE_TRX_FINISHED = 3
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:25]:
```suggestion
    SAVE_TXN_FINISHED = 3
```

File: src/sentry/utils/event_tracker.py [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+
+
+class TransactionStageStatus(IntEnum):
+    """
+    Do NOT change the int values of existing items. It will change meaning of existing tracking data
+    This class is DELETE or APPEND ONLY
+    """
+
+    # the transaction is stored to rc-transactions
+    REDIS_PUT = 1
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:19]: is there a reason for making these ints rather than strings or something else? are there size concerns?

personally i feel like this makes this code a lot less flexible and less easy to evolve

what if another stage between redis_put and save gets added later? will that then become 8? that's seems confusing.

File: src/sentry/utils/event_tracker.py [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+    POST_PROCESS_STARTED = 6
+
+    # the transaction is deleted from rc-transactions
+    REDIS_DELETED = 7
+
+
+logger = logging.getLogger("EventTracker")
+
+
+def should_track(event_id: str, sample_rate) -> bool:
```
Comment (lynnagara) [inline src/sentry/utils/event_tracker.py:43]: it's wasn't very clear at a glance that this is an internally used function only

```suggestion
def _should_track(event_id: str, sample_rate) -> bool:
```

Comment (lynnagara) [reply src/sentry/utils/event_tracker.py:43]: you could probably just inline this into `track_sampled_event`

File: src/sentry/utils/event_tracker.py (hunk 22 above)
Comment (evanpurkhiser) [reply src/sentry/utils/event_tracker.py:19]: 👍 for StrEnum for these, especially if they're being serialized in any way

Comment (fpacifici) [reply src/sentry/utils/event_tracker.py:19]: The reason originally was because it was planned to be stored in redis. smaller enums means higher sampling rate for the same memory usage. 
As we are not storing them in redis, there is no reason for int enum. Please turn them to strings.

File: src/sentry/utils/event_tracker.py [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def track_sampled_event(
+    event_id: str, event_type: str, status: TransactionStageStatus, sample_rate: float = None
+) -> None:
+    """
+    Records how far an event has made it through the ingestion pipeline.
+    Each event type will pick up its sampling rate from its registered option.
+    """
+    if sample_rate == 0:
```
Comment (fpacifici) [inline src/sentry/utils/event_tracker.py:61]: I think there is a misunderstanding. 
When I said to check the sampling rate to be 0 and return I meant read the sample rate option and check that. Please do not pass the sample rate to the method. THe option should be the only source of truth

File: src/sentry/options/defaults.py [hunk 25]
```diff
@@ -2897,3 +2897,4 @@
     default=[],
     flags=FLAG_ALLOW_EMPTY | FLAG_AUTOMATOR_MODIFIABLE,
 )
+register("performance.event-tracker.sample-rate.transaction", default=0.01)
```
[the remark this answers is not in the document]
Comment (victoria-yining-huang) [reply src/sentry/options/defaults.py]: done

**Comment by victoria-yining-huang** [comment]
> The reason originally was because it was planned to be stored in redis. smaller enums means higher sampling rate for the same memory usage.
> As we are not storing them in redis, there is no reason for int enum. Please turn them to strings.
@fpacifici int in logging should still be cheaper than string in google logs? Or is that too negligible

File: src/sentry/utils/event_tracker.py (hunk 24 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py:50]: ohhh understood, done

File: src/sentry/utils/event_tracker.py (hunk 23 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: done

File: src/sentry/ingest/consumer/processors.py (hunk 19 above)
Comment (victoria-yining-huang) [reply src/sentry/ingest/consumer/processors.py:207]: im using `data.get("type")` to make it generalized, so when this is eventually extended for errors, it will work too. Do you have any concerns a bout this?

Comment (victoria-yining-huang) [reply src/sentry/ingest/consumer/processors.py]: hardcoded it to only take transactions for now

File: src/sentry/utils/event_tracker.py (hunk 22 above)
Comment (victoria-yining-huang) [reply src/sentry/utils/event_tracker.py]: changed to strings

File: src/sentry/ingest/consumer/processors.py (hunk 19 above)
Comment (lynnagara) [reply src/sentry/ingest/consumer/processors.py]: Yes, you cannot use `data.get("type")` in errors - there are many different types going through the errors pipeline. The pipeline name should be the generalized thing.

File: tests/sentry/utils/test_event_tracker.py [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+
+    @patch("sentry.utils.event_tracker._do_record")
+    def test_track_sampled_event_does_not_log_event(self, mock_do_record):
+        with override_options({"performance.event-tracker.sample-rate.transaction": 0.0}):
+            track_sampled_event(EVENT_ID, EVENT_TYPE_STR, STATUS)
+            mock_do_record.assert_not_called()
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (lynnagara) [inline tests/sentry/utils/test_event_tracker.py:33]: why was this needed?
